src/Http/ResponseEmitter.php

```diff
@@ -213,68 +214,47 @@ protected function emitHeaders(ResponseInterface $response): void
     /**
      * Emit cookies using setcookie()
      *
-     * @param array $cookies An array of Set-Cookie headers.
+     * @param (string|\Cake\Http\Cookie\CookieInterface)[] $cookies An array of cookies.
      * @return void
      */
     protected function emitCookies(array $cookies): void
     {
         foreach ($cookies as $cookie) {
-            if (is_array($cookie)) {
-                setcookie(
-                    $cookie['name'],
-                    $cookie['value'],
-                    $cookie['expire'],
-                    $cookie['path'],
-                    $cookie['domain'],
-                    $cookie['secure'],
-                    $cookie['httpOnly']
-                );
-                continue;
-            }
+            $this->setCookie($cookie);
+        }
+    }
 
-            if (strpos($cookie, '";"') !== false) {
-                $cookie = str_replace('";"', '{__cookie_replace__}', $cookie);
-                $parts = str_replace('{__cookie_replace__}', '";"', explode(';', $cookie));
-            } else {
-                $parts = preg_split('/\;[ \t]*/', $cookie);
+    /**
+     * Helper methods to set cookie.
+     *
+     * For PHP 7.3 it uses setcookie() and for PHP < 7.3 it uses header().
+     *
+     * @param string|\Cake\Http\Cookie\CookieInterface $cookie Cookie.
+     * @return bool
+     */
+    protected function setCookie($cookie): bool
+    {
+        if ($cookie instanceof CookieInterface) {
+            if (PHP_VERSION_ID >= 70300) {
+                return setcookie($cookie->getName(), $cookie->getScalarValue(), $cookie->getOptions());
+            } elseif ($cookie->getSameSite() === null) {
+                return setcookie(
+                    $cookie->getName(),
+                    $cookie->getScalarValue(),
+                    $cookie->getExpiresTimestamp() ?: 0,
+                    $cookie->getPath(),
+                    $cookie->getDomain(),
+                    $cookie->isSecure(),
+                    $cookie->isHttpOnly()
+                );
             }
 
-            [$name, $value] = explode('=', array_shift($parts), 2);
-            $data = [
-                'name' => urldecode($name),
-                'value' => urldecode($value),
-                'expires' => 0,
-                'path' => '',
-                'domain' => '',
-                'secure' => false,
-                'httponly' => false,
-            ];
+            $cookie = $cookie->toHeaderValue();
+        }
 
-            foreach ($parts as $part) {
-                if (strpos($part, '=') !== false) {
-                    [$key, $value] = explode('=', $part);
-                } else {
-                    $key = $part;
-                    $value = true;
-                }
+        header('Set-Cookie: ' . $cookie);
```

Comment:
Have you checked that this will play nicely with sessions and setcookie()?

How will this work in integration tests?

Reply:
You need to pass in `false` as the second argument to allow multiple `Set-Cookie` headers.

Reply:
> Have you checked that this will play nicely with sessions and setcookie()?

IDK, I just followed @garas's suggestion. If it doesn't work then users of PHP 7.3 just won't be able to use `SameSite`.

> How will this work in integration tests?

There are no test failures :) 

> You need to pass in false as the second argument to allow multiple Set-Cookie headers.

Fixed

Reply:
Seems the `SameSite` restriction for session cookie can be handled this way https://github.com/delight-im/PHP-Cookie/blob/master/src/Session.php#L146.

Reply:
I'm not concerned about SameSite session cookies, more that using the `set-cookie` header can cause sessions to not work. This commit 7ebc2614fd9e32beec80a2c5dee9cf73f288bb1e has a removal in the `ResponseTransformer` that was a workaround for session cookies not being emitted.

Reply:
Okay, I really don't want to go down that rabbit hole for now. I'll revert the use of `header()` and simply throw exception is samesite is set for php 7.2.

We can look into supporting samesite attribute for PHP 7.2 at a later date.